src/wmc.rs

```diff
@@ -0,0 +1,566 @@
+// New Java switches are not correctly recognised by RCA, which returns some error nodes.
+// The presence of error nodes may cause wrong WMC measures.
+// However the issue has already been addressed and solved here:
+// https://github.com/tree-sitter/tree-sitter-java/pull/78
+
+use serde::Serialize;
+use std::path::Path;
+use std::path::PathBuf;
+use std::vec::Vec;
+
+use crate::cyclomatic::{self, Cyclomatic};
+use crate::node::Node;
+use crate::traits::*;
+use crate::*;
+
+#[derive(Default, Clone, Serialize)]
+pub struct Method {
+    pub name: String,
+    pub cc: f64,
+    pub classes: Vec<Class>,
+}
+
+#[derive(Default, Clone, Serialize)]
+pub struct Class {
+    pub name: String,
+    pub wmc: f64,
+    pub methods: Vec<Method>,
+    pub classes: Vec<Class>,
+}
+
+#[derive(Default, Serialize)]
+pub struct FileClasses {
+    pub name: String,
+    pub wmc_total: f64,
+    pub classes: Vec<Class>,
+}
+
+enum State {
+    Method(Method, cyclomatic::Stats),
+    Class(Class, cyclomatic::Stats),
+    File(FileClasses),
+}
+
+fn merge_wmc(child_state: &mut State, parent_state: &mut State) {
+    match child_state {
+        State::Method(m, m_stats) => match parent_state {
+            State::Class(c, c_stats) => {
+                m_stats.compute_sum();
+                c_stats.merge_sum(m_stats);
+
+                m.cc = m_stats.cyclomatic_sum();
+                c.wmc = c_stats.cyclomatic_sum();
+                c.methods.push(m.clone());
+            }
+            State::Method(outer_m, outer_m_stats) => {
+                m_stats.compute_sum();
+                outer_m_stats.merge_sum(m_stats);
+
+                outer_m.cc = outer_m_stats.cyclomatic_sum();
+            }
+            _ => {}
+        },
+        State::Class(c, c_stats) => match parent_state {
+            State::Method(m, m_stats) => {
+                c_stats.compute_sum();
+                m_stats.merge_sum(c_stats);
+
+                m.cc = m_stats.cyclomatic_sum();
+                m.classes.push(c.clone());
+            }
+            State::Class(outer_c, _) => {
+                outer_c.classes.push(c.clone());
+            }
+            State::File(f) => {
+                f.wmc_total += c.wmc;
+                f.classes.push(c.clone());
+            }
+        },
+        _ => {}
+    }
+}
+
+fn merge_states(state_stack: &mut Vec<State>, diff_level: usize) {
+    if state_stack.is_empty() {
+        return;
+    }
+    for _ in 0..diff_level {
+        if state_stack.len() == 1 {
+            break;
+        } else {
+            let child_state = &mut state_stack.pop().unwrap();
+            let parent_state = state_stack.last_mut().unwrap();
+            merge_wmc(child_state, parent_state);
+        }
+    }
+}
+
+pub fn wmc_metrics<T: ParserTrait>(parser: &T, path: &Path) -> Option<FileClasses> {
```

Comment:
I've the impress that it's a modified copy/paste of the code in `spaces.rs::metrics`...
Isn't it possible to avoid that ?
I think we must try to improve `spaces.rs::metrics` to fit your needs or maybe you must change your way to think to fit `spaces.rs::metrics`...
Duplicating code is rarely a good idea... so tbh I'm not super enthusiastic about this PR.
But of course if you've a very very good rationale to do that, we can discuss ;)

Reply:
Thank you very much for the code review and your comments.

The plan was to have an AST scan loop dedicated to all the object-oriented metrics computation.
In particular what I'm trying to do is to implement 3 of the 6 metrics defined in the CK metrics suite: WMC, DIT, and NOC.
WMC, DIT, and NOC are all metrics defined just for classes.
DIT (Depth of Inheritance Tree) and NOC (Number Of Children) are 2 metrics that focus on class inheritance.

WMC is quite similar to the Cyclomatic Complexity and I agree that it could be somehow implemented in `spaces.rs::metrics`.
However, to implement DIT and NOC I am next going to require a system to collect class data throughout different files before producing the DIT and NOC values for a single class.
Adding that system in metrics in my opinion will complicate a lot the function and will decrease a lot the performance.

Instead, I thought that having all 3 CK metrics using the same slimmer AST scan could simplify things:
1) I can use this faster AST scan to compute the object-oriented metrics that need to collect data from different files to produce values faster.
2) I can have all the 3 CK metrics computed in a simpler dedicated space so one can experiment more with them without any risks.
3) I can avoid printing out the WMC, DIT, and NOC values for spaces that are not classes since those metrics are not defined for non-class spaces.

Thank you for the opportunity to clarify and motivate my work.
Any additional advice and comment would be greatly appreciated and would help a lot to turn my work in the right direction.